src/hash.cr

```diff
@@ -804,113 +806,317 @@ class Hash(K, V)
     self
   end
 
-  def rehash
-    new_size = calculate_new_size(@size)
-    @buckets = @buckets.realloc(new_size)
-    new_size.times { |i| @buckets[i] = nil }
-    @buckets_size = new_size
-    entry = @last
-    while entry
-      index = bucket_index entry.key
-      entry.next = @buckets[index]
-      @buckets[index] = entry
-      entry = entry.back
-    end
-  end
-
   # Inverts keys and values. If there are duplicated values, the last key becomes the new value.
   #
   # ```
   # {"foo" => "bar"}.invert                 # => {"bar" => "foo"}
   # {"foo" => "bar", "baz" => "bar"}.invert # => {"bar" => "baz"}
   # ```
   def invert
-    hash = Hash(V, K).new(initial_capacity: @buckets_size)
+    hash = Hash(V, K).new(initial_capacity: @size)
     self.each do |k, v|
       hash[v] = k
     end
     hash
   end
 
-  protected def find_entry(key)
-    return nil if empty?
+  # Implementation
 
-    index = bucket_index key
-    entry = @buckets[index]
-    find_entry_in_bucket entry, key
+  CHUNK = 8_u32
+
+  @[AlwaysInline]
+  private def chunks_ptr
+    capa = nentries(@format)
+    if capa <= CHUNK
+      pointerof(@entries).as(Pointer(Pointer(Entry(K, V))))
+    else
+      @entries.as(Pointer(Pointer(Entry(K, V))))
+    end
   end
 
-  private def insert_in_bucket(index, key, value)
-    entry = @buckets[index]
-    if entry
-      while entry
-        if entry.key == key
-          entry.value = value
-          return nil
-        end
-        if entry.next
-          entry = entry.next
-        else
-          return entry.next = Entry(K, V).new(key, value)
-        end
+  protected def entry_at(i)
+    chunks = chunks_ptr
+    chunks[i / CHUNK] + i % CHUNK
+  end
+
+  @[AlwaysInline]
+  private def each_entry
+    return if empty?
+    rnum = @rebuild_num
+    chunks = chunks_ptr
+    @first.upto(@last - 1) do |i|
+      chunk = chunks[i / CHUNK]
+      entry = chunk + i % CHUNK
```

Comment:
I'm not certain there is a noticeable difference or if llvm is smart enough to notice this by itself, but perhaps 
```
chunk_index, offset = i.divmod(CHUNK)
```

Reply:
llvm should optimize division and module by 8 to shifts and mask (ie `i >> 3` and `i & 7`)